### DISCUSSION TITLE
Fastest provider

### DISCUSSION THREAD

**Discussion by automartin5000** [CloudSnorkel/cdk-github-runners#260] [Q&A] [answered]
I started using CodeBuild as the provider as it seemed like the simplest solution given that its serverless. But it seems like provisioning times are at least 1 minute, presumably to pull the docker image. Is there a way to speed that up?

So I tried EC2 spot, figuring that'd be cheap and always available, but it seems as though it doesn't leave the instances running and spins up a new instance for every job. Is there a way to leave `n` number of instances running at a max price?

All this is to say: what's the fastest provider?

**Comment by kichik** [comment]
Lambda is the fastest provider but it is very limited by having a read-only file system, short time limit, and only 10GB of space. CodeBuild can be sped up by using bigger instances. EC2 might be faster still.

None of the options leave the runners behind because we aim to spin up ephemeral runners. Persistent runners can cause weird build issues as anything installed in one build will be available in the next. 

It sounds like you're looking for #191.

**Reply by automartin5000** [reply]
I’m still seeing one minute initialization times with Lambda (presumably to pull the image). It seems like every option has about a 1 minute start to pull the image. Even more reason for #191

**Reply by kichik** [reply, marked as answer]
0.9.1 now has support for `EcsRunnerProvider` and you can use something like:

```
new GitHubRunners(stack, 'runners', {
  providers: [
    new EcsRunnerProvider(stack, 'ECS', {
      minInstances: 1,
  ],
});
```

Even with the default `minInstances: 0`, the instances stay behind for 15 minutes. You have access to the ASG with `new EcsRunnerProvider(this, 'ECS').capacityProvider.autoScalingGroup` so you should be able to even add scheduled scaling or anything else you might need. And if you want even more control, you can override the capacity provider completely.

**Reply by automartin5000** [reply]
Awesome thanks! Can’t wait to try it 💪

**Reply by sercantor** [reply]
@kichik if I use ECS Runners, and set the `minInstances: 0` will the runners still be ephemeral?

**Reply by kichik** [reply]
> @kichik if I use ECS Runners, and set the `minInstances: 0` will the runners still be ephemeral?

Yes. The Docker containers running the runner are never reused.

**Reply by sercantor** [reply]
Thanks a lot for answering, I didn't know the inner workings 🙏
